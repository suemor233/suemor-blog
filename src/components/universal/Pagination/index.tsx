import clsx from 'clsx'
import { useRouter } from 'next/router'
import type { FC, HTMLProps, PropsWithChildren } from 'react'
import { memo, useEffect, useMemo, useState } from 'react'
import {
  MdOutlineArrowBackIosNew,
  MdOutlineArrowForwardIos,
} from 'react-icons/md'

export interface PaginationProps {
  total: number
  initialPage: number
  onChange?: (page: number) => void
}

const Pagination: FC<PaginationProps> = (props) => {
  const { total, initialPage } = props
  const [currentPage, setCurrentPage] = useState(initialPage)
  const router = useRouter()
  useEffect(() => {
    !router?.query.page && setCurrentPage(initialPage)
  }, [router?.query])
  const paginationCenter = useMemo(() => {
    const front = () => {
      const arr: number[] = []
      for (let i = 2; i > 0; i--) {
        if (currentPage > 3) {
          arr.push(currentPage - i)
        } else {
          if (currentPage === 3) {
            arr.push(2)
          }
          break
        }
      }
      return arr
    }

    const center = () => {
      const arr: number[] = []
      if (!(currentPage === 1 || currentPage === total)) {
        arr.push(currentPage)
      }
      return arr
    }

    const behind = () => {
      const arr: number[] = []
      for (let i = 1; i < 3; i++) {
        if (total - currentPage > 2) {
          arr.push(currentPage + i)
        } else {
          if (total - currentPage === 2) {
            arr.push(total - 1)
          }
          break
        }
      }
      return arr
    }
    return [
      total != 1 ? 1 : [],
      currentPage > 4 ? '...' : [],
      ...front(),
      ...center(),
      ...behind(),
      total - 3 > currentPage ? '...' : [],
      total,
    ].flat()
  }, [currentPage, total])

  const handleCurrentPage = (num: number) => {
    if (num < 1 || num > total) {
      return
    }
    setCurrentPage(num)
    props.onChange && props.onChange(num)
  }

  return (
    <section className="flex gap-2">
      {currentPage !== 1 && (
        <Button onClick={() => handleCurrentPage(currentPage - 1)}>
          <MdOutlineArrowBackIosNew />
        </Button>
      )}

      {paginationCenter.map((item, index) => (
        <Button
          key={index}
          onClick={() =>
            typeof item !== 'string' && item !== currentPage
              ? handleCurrentPage(item)
              : null
          }
          className={clsx(
            item === currentPage && '!bg-blue-500 text-white hover:bg-blue-500',
          )}
        >
          {item}
        </Button>
      ))}
      {![total].includes(currentPage) && (
        <Button onClick={() => handleCurrentPage(currentPage + 1)}>
          <MdOutlineArrowForwardIos />
        </Button>
      )}
    </section>
  )
}

const Button: FC<PropsWithChildren & HTMLProps<HTMLButtonElement>> = ({
  children,
  onClick,
  className,
}) => {
  return (
    <button
      className={clsx(
        'bg-gray-100 h-9 w-9 rounded-md hover:bg-gray-200 flex justify-center items-center cursor-pointer dark:bg-black',
        className,
      )}
      onClick={onClick}
      aria-label="pagination"
    >
      {children}
    </button>
  )
}

export default memo(Pagination)
