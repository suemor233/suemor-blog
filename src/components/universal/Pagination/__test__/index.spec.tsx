// import { fireEvent, render } from '@testing-library/react'
import React from 'React'
import { describe, expect, it } from 'vitest'

import { render, screen ,fireEvent} from '@testing-library/react'

import Pagination from '..'

const PaginationProps = {
  total: 10,
  initialPage: 1,
}
describe('Pagination', () => {
  it('pagination should be init correctly', () => {
    render(<Pagination {...PaginationProps} />)
    const selectPagination = document.querySelector('.text-white')
    const totalPagination = screen.getAllByRole('button', {
      name: 'pagination',
    })

    expect(selectPagination?.innerHTML).toBe(`${PaginationProps.initialPage}`)
    expect(totalPagination[4].innerHTML).toBe(`${PaginationProps.total}`)
    expect(selectPagination).toMatchSnapshot()
  })

  it('pagination should be unmount correctly', () => {
    const component = render(<Pagination {...PaginationProps} />)
    expect(() => {
      component.unmount()
    }).not.toThrow()
  })

  it('pagination should jump correctly pages',()=> {
    render(<Pagination {...PaginationProps} />)
    const totalPagination = screen.getAllByRole('button', {
      name: 'pagination',
    })
    fireEvent.click(totalPagination[2])
    expect(totalPagination[2].innerHTML).toBe('3')
  })

  it('pagination snapshot', () => {
    const component = render(<Pagination {...PaginationProps} />)
    expect(component).toMatchSnapshot()
  })
})
