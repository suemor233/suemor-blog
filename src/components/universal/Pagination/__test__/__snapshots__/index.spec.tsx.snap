// Vitest Snapshot v1

exports[`Pagination > pagination should be init correctly 1`] = `
<button
  aria-label="pagination"
  class="bg-gray-100 h-9 w-9 rounded-md hover:bg-gray-200 flex justify-center items-center cursor-pointer dark:bg-black !bg-blue-500 text-white hover:bg-blue-500"
>
  1
</button>
`;

exports[`Pagination > pagination snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        class="flex gap-2"
      >
        <button
          aria-label="pagination"
          class="bg-gray-100 h-9 w-9 rounded-md hover:bg-gray-200 flex justify-center items-center cursor-pointer dark:bg-black"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M17.77 3.77L16 2 6 12l10 10 1.77-1.77L9.54 12z"
            />
          </svg>
        </button>
        <button
          aria-label="pagination"
          class="bg-gray-100 h-9 w-9 rounded-md hover:bg-gray-200 flex justify-center items-center cursor-pointer dark:bg-black"
        >
          1
        </button>
        <button
          aria-label="pagination"
          class="bg-gray-100 h-9 w-9 rounded-md hover:bg-gray-200 flex justify-center items-center cursor-pointer dark:bg-black"
        >
          2
        </button>
        <button
          aria-label="pagination"
          class="bg-gray-100 h-9 w-9 rounded-md hover:bg-gray-200 flex justify-center items-center cursor-pointer dark:bg-black !bg-blue-500 text-white hover:bg-blue-500"
        >
          3
        </button>
        <button
          aria-label="pagination"
          class="bg-gray-100 h-9 w-9 rounded-md hover:bg-gray-200 flex justify-center items-center cursor-pointer dark:bg-black"
        >
          4
        </button>
        <button
          aria-label="pagination"
          class="bg-gray-100 h-9 w-9 rounded-md hover:bg-gray-200 flex justify-center items-center cursor-pointer dark:bg-black"
        >
          5
        </button>
        <button
          aria-label="pagination"
          class="bg-gray-100 h-9 w-9 rounded-md hover:bg-gray-200 flex justify-center items-center cursor-pointer dark:bg-black"
        >
          ...
        </button>
        <button
          aria-label="pagination"
          class="bg-gray-100 h-9 w-9 rounded-md hover:bg-gray-200 flex justify-center items-center cursor-pointer dark:bg-black"
        >
          10
        </button>
        <button
          aria-label="pagination"
          class="bg-gray-100 h-9 w-9 rounded-md hover:bg-gray-200 flex justify-center items-center cursor-pointer dark:bg-black"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0V0z"
              fill="none"
            />
            <path
              d="M6.23 20.23L8 22l10-10L8 2 6.23 3.77 14.46 12z"
            />
          </svg>
        </button>
      </section>
    </div>
    <div />
    <div>
      <section
        class="flex gap-2"
      >
        <button
          aria-label="pagination"
          class="bg-gray-100 h-9 w-9 rounded-md hover:bg-gray-200 flex justify-center items-center cursor-pointer dark:bg-black !bg-blue-500 text-white hover:bg-blue-500"
        >
          1
        </button>
        <button
          aria-label="pagination"
          class="bg-gray-100 h-9 w-9 rounded-md hover:bg-gray-200 flex justify-center items-center cursor-pointer dark:bg-black"
        >
          2
        </button>
        <button
          aria-label="pagination"
          class="bg-gray-100 h-9 w-9 rounded-md hover:bg-gray-200 flex justify-center items-center cursor-pointer dark:bg-black"
        >
          3
        </button>
        <button
          aria-label="pagination"
          class="bg-gray-100 h-9 w-9 rounded-md hover:bg-gray-200 flex justify-center items-center cursor-pointer dark:bg-black"
        >
          ...
        </button>
        <button
          aria-label="pagination"
          class="bg-gray-100 h-9 w-9 rounded-md hover:bg-gray-200 flex justify-center items-center cursor-pointer dark:bg-black"
        >
          10
        </button>
        <button
          aria-label="pagination"
          class="bg-gray-100 h-9 w-9 rounded-md hover:bg-gray-200 flex justify-center items-center cursor-pointer dark:bg-black"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0V0z"
              fill="none"
            />
            <path
              d="M6.23 20.23L8 22l10-10L8 2 6.23 3.77 14.46 12z"
            />
          </svg>
        </button>
      </section>
    </div>
    <div>
      <section
        class="flex gap-2"
      >
        <button
          aria-label="pagination"
          class="bg-gray-100 h-9 w-9 rounded-md hover:bg-gray-200 flex justify-center items-center cursor-pointer dark:bg-black !bg-blue-500 text-white hover:bg-blue-500"
        >
          1
        </button>
        <button
          aria-label="pagination"
          class="bg-gray-100 h-9 w-9 rounded-md hover:bg-gray-200 flex justify-center items-center cursor-pointer dark:bg-black"
        >
          2
        </button>
        <button
          aria-label="pagination"
          class="bg-gray-100 h-9 w-9 rounded-md hover:bg-gray-200 flex justify-center items-center cursor-pointer dark:bg-black"
        >
          3
        </button>
        <button
          aria-label="pagination"
          class="bg-gray-100 h-9 w-9 rounded-md hover:bg-gray-200 flex justify-center items-center cursor-pointer dark:bg-black"
        >
          ...
        </button>
        <button
          aria-label="pagination"
          class="bg-gray-100 h-9 w-9 rounded-md hover:bg-gray-200 flex justify-center items-center cursor-pointer dark:bg-black"
        >
          10
        </button>
        <button
          aria-label="pagination"
          class="bg-gray-100 h-9 w-9 rounded-md hover:bg-gray-200 flex justify-center items-center cursor-pointer dark:bg-black"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0V0z"
              fill="none"
            />
            <path
              d="M6.23 20.23L8 22l10-10L8 2 6.23 3.77 14.46 12z"
            />
          </svg>
        </button>
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      class="flex gap-2"
    >
      <button
        aria-label="pagination"
        class="bg-gray-100 h-9 w-9 rounded-md hover:bg-gray-200 flex justify-center items-center cursor-pointer dark:bg-black !bg-blue-500 text-white hover:bg-blue-500"
      >
        1
      </button>
      <button
        aria-label="pagination"
        class="bg-gray-100 h-9 w-9 rounded-md hover:bg-gray-200 flex justify-center items-center cursor-pointer dark:bg-black"
      >
        2
      </button>
      <button
        aria-label="pagination"
        class="bg-gray-100 h-9 w-9 rounded-md hover:bg-gray-200 flex justify-center items-center cursor-pointer dark:bg-black"
      >
        3
      </button>
      <button
        aria-label="pagination"
        class="bg-gray-100 h-9 w-9 rounded-md hover:bg-gray-200 flex justify-center items-center cursor-pointer dark:bg-black"
      >
        ...
      </button>
      <button
        aria-label="pagination"
        class="bg-gray-100 h-9 w-9 rounded-md hover:bg-gray-200 flex justify-center items-center cursor-pointer dark:bg-black"
      >
        10
      </button>
      <button
        aria-label="pagination"
        class="bg-gray-100 h-9 w-9 rounded-md hover:bg-gray-200 flex justify-center items-center cursor-pointer dark:bg-black"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0V0z"
            fill="none"
          />
          <path
            d="M6.23 20.23L8 22l10-10L8 2 6.23 3.77 14.46 12z"
          />
        </svg>
      </button>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
